javalib/src/main/scala/java/util/PriorityQueue.scala

```diff
@@ -47,47 +62,82 @@ class PriorityQueue[E] private (
         NaturalComparator.select(c.comparator().asInstanceOf[Comparator[_ >: E]])
       case _ =>
         NaturalComparator
-    }, internal = true)
+    }, internal = true, nextPowerOfTwo(c.size()))
     addAll(c)
   }
 
   def this(c: PriorityQueue[_ <: E]) = {
-    this(c.comp.asInstanceOf[Comparator[_ >: E]], internal = true)
+    this(c.comp.asInstanceOf[Comparator[_ >: E]], internal = true, nextPowerOfTwo(c.size()))
     addAll(c)
   }
 
   def this(sortedSet: SortedSet[_ <: E]) = {
     this(NaturalComparator.select(
         sortedSet.comparator().asInstanceOf[Comparator[_ >: E]]),
-        internal = true)
+        internal = true,
+        nextPowerOfTwo(sortedSet.size()))
     addAll(sortedSet)
   }
 
+  /* This class has two different implementations for the internal storage,
+   * depending on whether we are on Wasm or JS.
+   * On JS, we utilize `js.Array`. On Wasm, for performance reasons,
+   * we use a scala.Array to avoid JS interop.
+   */
+
   // The index 0 is not used; the root is at index 1.
   // This is standard practice in binary heaps, to simplify arithmetics.
-  private[this] val inner = js.Array[E](null.asInstanceOf[E])
+  private val innerJS: js.Array[E] =
+    if (LinkingInfo.isWebAssembly) null
+    else js.Array[E](null.asInstanceOf[E])
+
+  private var innerWasm: Array[AnyRef] =
+    if (LinkingInfo.isWebAssembly) new Array[AnyRef](initialCapacity)
+    else null
+
+  // Wasm only: size of the objects stored in the inner array
+  private var innerSize = 1
 
   override def add(e: E): Boolean = {
     if (e == null)
       throw new NullPointerException()
-    inner.push(e)
-    fixUp(inner.length - 1)
+    if (LinkingInfo.isWebAssembly) {
+      ensureCapacity(innerSize + 1)
+      innerWasm(innerSize) = e.asInstanceOf[AnyRef]
+      innerSize += 1
+    } else {
+      innerJS.push(e)
+    }
+    fixUp(length() - 1)
     true
   }
 
   def offer(e: E): Boolean = add(e)
 
-  def peek(): E =
-    if (inner.length > 1) inner(1)
-    else null.asInstanceOf[E]
+  def peek(): E = {
+    if (LinkingInfo.isWebAssembly) {
+      if (innerSize > 1) innerWasm(1).asInstanceOf[E]
+      else null.asInstanceOf[E]
```

Comment:
I think a *lot* could be simplified in `PriorityQueue` with 3 internal helpers: `innerLength`, `innerGet` and `innerSet`. This would free all the rest from altering their logic.